File: src/weaponsprocurement/gui/StockReviewSourceState.java
package weaponsprocurement.gui;

import weaponsprocurement.core.StockReviewConfig;
import weaponsprocurement.core.StockSortMode;
import weaponsprocurement.core.StockSourceMode;

final class StockReviewSourceState {
    private StockSortMode sortMode;
    private boolean includeCurrentMarketStorage;
    private boolean includeBlackMarket;
    private StockSourceMode sourceMode;

    StockReviewSourceState(StockReviewConfig config) {
        this.sortMode = config.getSortMode();
        this.includeCurrentMarketStorage = config.isIncludeCurrentMarketStorage();
        this.includeBlackMarket = config.isIncludeBlackMarket();
        this.sourceMode = StockSourceMode.LOCAL;
    }

    StockReviewSourceState(StockReviewSourceState source) {
        this.sortMode = source.sortMode;
        this.includeCurrentMarketStorage = source.includeCurrentMarketStorage;
        this.includeBlackMarket = source.includeBlackMarket;
        this.sourceMode = source.sourceMode;
    }

    StockSortMode getSortMode() {
        return sortMode;
    }

    void cycleSortMode() {
        sortMode = sortMode.next();
    }

    boolean isIncludeCurrentMarketStorage() {
        return includeCurrentMarketStorage;
    }

    boolean isIncludeBlackMarket() {
        return !getSourceMode().isRemote() && includeBlackMarket;
    }

    void toggleBlackMarket() {
        if (getSourceMode().isRemote()) {
            includeBlackMarket = false;
            return;
        }
        includeBlackMarket = !includeBlackMarket;
    }

    StockSourceMode getSourceMode() {
        StockSourceMode resolved = sourceMode == null ? StockSourceMode.LOCAL : sourceMode;
        if (!resolved.isEnabled()) {
            sourceMode = StockSourceMode.LOCAL;
            includeBlackMarket = false;
            return sourceMode;
        }
        return resolved;
    }

    void cycleSourceMode() {
        sourceMode = getSourceMode().next();
        if (sourceMode.isRemote()) {
            includeBlackMarket = false;
        }
    }
}

File: src/main/kotlin/weaponsprocurement/gui/StockReviewSourceState.kt
package weaponsprocurement.gui

import weaponsprocurement.config.StockReviewConfig
import weaponsprocurement.stock.StockSortMode
import weaponsprocurement.stock.StockSourceMode

class StockReviewSourceState {
    private var sortMode: StockSortMode
    private var includeCurrentMarketStorage: Boolean
    private var includeBlackMarket: Boolean
    private var sourceMode: StockSourceMode?

    constructor(config: StockReviewConfig) {
        sortMode = config.getSortMode()
        includeCurrentMarketStorage = config.isIncludeCurrentMarketStorage()
        includeBlackMarket = config.isIncludeBlackMarket()
        sourceMode = StockSourceMode.LOCAL
    }

    constructor(source: StockReviewSourceState) {
        sortMode = source.sortMode
        includeCurrentMarketStorage = source.includeCurrentMarketStorage
        includeBlackMarket = source.includeBlackMarket
        sourceMode = source.sourceMode
    }

    fun getSortMode(): StockSortMode = sortMode

    fun cycleSortMode() {
        sortMode = sortMode.next()
    }

    fun isIncludeCurrentMarketStorage(): Boolean = includeCurrentMarketStorage

    fun isIncludeBlackMarket(): Boolean = !getSourceMode().isRemote() && includeBlackMarket

    fun toggleBlackMarket() {
        if (getSourceMode().isRemote()) {
            includeBlackMarket = false
            return
        }
        includeBlackMarket = !includeBlackMarket
    }

    fun getSourceMode(): StockSourceMode {
        val resolved = sourceMode ?: StockSourceMode.LOCAL
        if (!resolved.isEnabled) {
            sourceMode = StockSourceMode.LOCAL
            includeBlackMarket = false
            return sourceMode as StockSourceMode
        }
        return resolved
    }

    fun cycleSourceMode() {
        sourceMode = getSourceMode().next()
        if (sourceMode!!.isRemote()) {
            includeBlackMarket = false
        }
    }
}
